### PR TITLE
8245145: Spinner: throws IllegalArgumentException when replacing skin

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Fixed SpinnerSkin initialization using new Skin.install().  Also in this PR - fixing memory leaks in SpinnerSkin by properly removing all listeners and event filters added in the constructor, as well as any child Nodes.

NOTE: the fix requires both ListenerHelper [JDK-8294809](https://bugs.openjdk.org/browse/JDK-8294809) and Skin.install() [JDK-8290844](https://bugs.openjdk.org/browse/JDK-8290844) changes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8245145](https://bugs.openjdk.org/browse/JDK-8245145): Spinner: throws IllegalArgumentException when replacing skin

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/904/head:pull/904` \
`$ git checkout pull/904`

Update a local copy of the PR: \
`$ git checkout pull/904` \
`$ git pull https://git.openjdk.org/jfx pull/904/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 904`

View PR using the GUI difftool: \
`$ git pr show -t 904`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/904.diff">https://git.openjdk.org/jfx/pull/904.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/SpinnerSkin.java [hunk 1]
```diff
@@ -261,13 +262,25 @@ private boolean isIncDecKeyEvent(KeyEvent ke) {
      *                                                                         *
      **************************************************************************/
 
+    @Override
+    public void install() {
+        getChildren().add(textField);
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/SpinnerSkin.java:267]: `dispose()` method removes 3 children, whereas `install()` method adds only 1 child. Is this intentional?

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/SpinnerSkin.java:269]: good question!  yes, it is intentional - the textField, unlike the buttons, comes from the control, so when we change skins the old skin removes it in its dispose(), followed by the new skin adding it back in install(), as it cannot do that in the constructor.

added a comment in the code to clarify

**Comment by andy-goryachev-oracle** [comment]
@nlisker or @kevinrushforth could you take a look please?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/SpinnerSkin.java [hunk 2]
```diff
@@ -261,13 +262,27 @@ private boolean isIncDecKeyEvent(KeyEvent ke) {
...
+        // by the old skin in its dispose(), followed by (re-)adding it here.
+        getChildren().add(textField);
+    }
+
     /** {@inheritDoc} */
-    @Override public void dispose() {
-        super.dispose();
+    @Override
+    public void dispose() {
+        if(getSkinnable() == null) {
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/SpinnerSkin.java:275]: Minor: add a space after the `if`.

**Review by kevinrushforth — APPROVED** [review]
Looks good with one minor formatting issue.

**Review by aghaisas — APPROVED** [review]

**Pull request closed without merging**